File: edc/org.eclipse.cdt.debug.edc.windows.agent/src/win_agent/ContextManager.cpp
/*******************************************************************************
 * Copyright (c) 2009 Nokia and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 * Nokia - Initial API and implementation
 *******************************************************************************/
#include "ContextManager.h"
#include "AgentUtils.h"

std::map<ContextID, Context*> ContextManager::gDebuggedContexts;
std::map<ContextID, Context*> ContextManager::gRunningContexts;


void ContextManager::addDebuggedContext(Context* context) {
	if (context != NULL)
		gDebuggedContexts[context->GetID()] = context;
}

void ContextManager::addRunningContext(Context* context) {
	if (context != NULL)
		gRunningContexts[context->GetID()] = context;
}

/*
 * return NULL pointer if not found.
 */
Context* ContextManager::findDebuggedContext(ContextID id) {
	// NOTE that if "id" is not in the map a, a[id] does not return NULL !
	// See STL Map document for more.
	std::map<ContextID, Context*>::iterator iter = gDebuggedContexts.find(id);
	if (iter != gDebuggedContexts.end())
		return gDebuggedContexts[id];
	else
		return NULL;
}

/*
 * return NULL pointer if not found.
 */
Context* ContextManager::findRunningContext(ContextID id) {
	// NOTE that if "id" is not in the map a, a[id] does not return NULL !
	// See STL Map document for more.
	std::map<ContextID, Context*>::iterator iter = gRunningContexts.find(id);
	if (iter != gRunningContexts.end())
		return gRunningContexts[id];
	else
		return NULL;
}

/*
 * Remove given context from the cache.
 * If the given context is not in the cache, this is a void operation.
 */
void ContextManager::removeDebuggedContext(ContextID id) {
	gDebuggedContexts.erase(id);
}

/**
 * Delete all cached contexts and flush the cache.
 */
void ContextManager::clearRunningContextCache() {
	std::map<ContextID, Context*>::iterator iter = gRunningContexts.begin();
	while (iter != gRunningContexts.end()) {
		delete iter->second;
		iter++;
	}

	gRunningContexts.clear();
}

/**
 * Delete all cached contexts and flush the cache.
 */
void ContextManager::clearDebuggedContextCache() {
	std::map<ContextID, Context*> copy = gDebuggedContexts;
	std::map<ContextID, Context*>::iterator iter = copy.begin();
	while (iter != copy.end()) {
		delete iter->second;
		iter++;
	}

	gDebuggedContexts.clear();
}

/**
 * Delete all cached contexts and flush the cache.
 */
void ContextManager::clearContextCache() {
	clearRunningContextCache();
	clearDebuggedContextCache();
}

File: edc/org.eclipse.cdt.debug.edc.tcf.extension/src/org/eclipse/cdt/debug/edc/tcf/extension/agent/ContextManager.java
/*******************************************************************************
 * Copyright (c) 2009, 2010 Nokia Corporation and/or its subsidiaries
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *    Ling Wang (Nokia) - initial version. Oct 19, 2009
 *******************************************************************************/

package org.eclipse.cdt.debug.edc.tcf.extension.agent;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Manager of all contexts (threads and processes, etc) in the debug session.
 */
public class ContextManager {
	/**
	 * All contexts under debug.
	 */
	private static Map<String, ContextInAgent> gDebuggedContexts = new HashMap<String, ContextInAgent>();
	
	/**
	 * Running contexts in the target OS that may or may not be under debug.
	 */
	private static Map<String, ContextInAgent> gRunningContexts = new HashMap<String, ContextInAgent>();
	
	static public void addDebuggedContext(ContextInAgent c) {
		assert c.getID() != null;	// internal ID has been given.
		gDebuggedContexts.put(c.getID(), c);
	}
	
	/**
	 * Remove given context from the debugged context cache.
	 * 
	 * @param id internal context ID of the context to remove.
	 * @param removeChildren whether to remove children of the context from the cache.
	 */
	static public void removeDebuggedContext(String id, boolean removeChildren) {
		if (removeChildren) {
			ContextInAgent cia = findDebuggedContext(id);
			if (cia != null)
			{
				for (String c : cia.getChildren()) {
					removeDebuggedContext(c, true);
				}				
			}
		}
		
		gDebuggedContexts.remove(id);
	}
	
	/**
	 * Get list of context IDs under debug.
	 * 
	 * @return a list of IDs of contexts under debug.
	 */
	static public List<String> getDebuggedContexts() {
		List<String> ret = new ArrayList<String>();
		for (ContextInAgent c : gDebuggedContexts.values())
			ret.add(c.getID());
		
		return ret;
	}
	
	/**
	 * Whether there are still context in the cache of debugged contexts.
	 * 
	 * @return
	 */
	static public boolean hasDebuggedContext() {
		return gDebuggedContexts.size() > 0;
	}
	
	/**
	 * Find the context with given ID.
	 * @param id  internal ID of the context
	 * @return the context found, null if not found.
	 */
	static public ContextInAgent findDebuggedContext(String id) {
		return gDebuggedContexts.get(id);
	}

	static public void addRunningContext(ContextInAgent c) {
		gRunningContexts.put(c.getID(), c);
	}

	static public void removeRunningContext(String id) {
		gRunningContexts.remove(id);
	}
	
	/**
	 * Find the running context with given ID.
	 * @param id  internal ID of the context
	 * @return the context found, null if not found.
	 */
	static public ContextInAgent findRunningContext(String id) {
		return gRunningContexts.get(id);
	}
	
	static public void clearRunningContextCache() {
		gRunningContexts.clear();
	}

	static public void clearDebuggedContextCache() {
		gDebuggedContexts.clear();
	}

	/**
	 * Clear all context caches.
	 */
	static public void clearContextCache() {
		clearRunningContextCache();
		clearDebuggedContextCache();
	}
}
